src/v/arch/manifest.cc

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "arch/manifest.h"
+
+#include "arch/error.h"
+#include "arch/logger.h"
+#include "bytes/iobuf.h"
+#include "bytes/iobuf_istreambuf.h"
+#include "bytes/iobuf_ostreambuf.h"
+#include "hashing/murmur.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "storage/ntp_config.h"
+
+#include <seastar/core/coroutine.hh>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+
+#include <algorithm>
+#include <array>
+
+namespace arch {
+
+manifest::manifest()
+  : _ntp()
+  , _rev() {}
+
+manifest::manifest(model::ntp ntp, model::revision_id rev)
+  : _ntp(std::move(ntp))
+  , _rev(rev) {}
+
+remote_manifest_path manifest::get_manifest_path() const {
+    constexpr uint32_t bitmask = 0xF0000000;
+    auto path = fmt::format("{}_{}", _ntp.path(), _rev());
+    uint32_t hash = bitmask & murmurhash3_x86_32(path.data(), path.size());
+    return remote_manifest_path(fmt::format(
+      "{:08x}/meta/{}_{}/manifest.json", hash, _ntp.path(), _rev()));
+}
+
+remote_segment_path
+manifest::get_remote_segment_path(const segment_name& name) const {
+    auto path = fmt::format("{}_{}/{}", _ntp.path(), _rev(), name());
+    uint32_t hash = murmurhash3_x86_32(path.data(), path.size());
+    return remote_segment_path(fmt::format("{:08x}/{}", hash, path));
+}
+
+model::ntp manifest::get_ntp() const { return _ntp; }
+
+model::revision_id manifest::get_revision_id() const { return _rev; }
+
+manifest::const_iterator manifest::begin() const { return _segments.begin(); }
+
+manifest::const_iterator manifest::end() const { return _segments.end(); }
+
+size_t manifest::size() const { return _segments.size(); }
+
+bool manifest::contains(const segment_name& obj) const {
+    return _segments.count(obj) != 0;
+}
+
+bool manifest::add(const segment_name& key, const segment_meta& meta) {
+    auto [it, ok] = _segments.insert(std::make_pair(key, meta));
+    return ok;
+}
+
+std::optional<std::reference_wrapper<const manifest::segment_meta>>
+manifest::get(const segment_name& key) const {
+    auto it = _segments.find(key);
+    if (it == _segments.end()) {
+        return std::nullopt;
+    }
+    return std::cref(it->second);
+}
+
+std::insert_iterator<manifest::segments_map> manifest::get_insert_iterator() {
+    return std::inserter(_segments, _segments.begin());
+}
+
+manifest manifest::difference(const manifest& remote_set) const {
+    if (_ntp != remote_set._ntp && _rev != remote_set._rev) {
+        throw manifest_error("NTPs doesn't match");
```

Comment:
typo: `ntps do not match` - thought i'd use this. https://github.com/vectorizedio/redpanda/blob/dev/src/v/vlog.h#L39